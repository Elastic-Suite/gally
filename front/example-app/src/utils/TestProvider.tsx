import { ReactNode } from 'react'

import { IApi, schemaContext } from 'shared'

interface IProps {
  api: IApi
  children: ReactNode
}

function TestProvider(props: IProps): JSX.Element {
  const { api, children } = props
  return <schemaContext.Provider value={api}>{children}</schemaContext.Provider>
}

export default TestProvider
