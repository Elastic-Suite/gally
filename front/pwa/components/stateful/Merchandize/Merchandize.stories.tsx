import { ComponentMeta, ComponentStory } from '@storybook/react'
import { useState } from 'react'

import Merchandize from './Merchandize'

export default {
  title: 'Stateful/Merchandize',
  component: Merchandize,
} as ComponentMeta<typeof Merchandize>

const Template: ComponentStory<typeof Merchandize> = (args): JSX.Element => {
  const [catConf, setCatConf] = useState({
    '@id': '/category_configurations/6',
    '@type': 'CategoryConfiguration',
    id: 6,
    category: 'cat',
    defaultSorting: '10',
    isActive: false,
    isVirtual: false,
    name: 'Category',
    useNameInProductSearch: true,
  })

  const handleChange = (name: string, value: boolean | string): void => {
    setCatConf({
      ...catConf,
      [name]: value,
    })
  }

  return <Merchandize catConf={catConf} onChange={handleChange} {...args} />
}

export const Default = Template.bind({})
Default.args = {
  sortOptions: [
    { label: 'Position', value: '10' },
    { label: 'Product Name', value: '20' },
    { label: 'Price', value: '30' },
    { label: 'Performance', value: '40' },
  ],
}
