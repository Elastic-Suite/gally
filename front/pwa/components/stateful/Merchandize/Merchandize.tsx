import { SyntheticEvent } from 'react'
import { useTranslation } from 'next-i18next'
import { Grid, Paper } from '@mui/material'
import {
  ICategory,
  IOptions,
  IParsedCategoryConfiguration,
  getFieldConfig,
  getFieldState,
  isVirtualCategoryEnabled,
} from 'shared'

import { useResource } from '~/hooks'
import { selectBundles, useAppSelector } from '~/store'

import DropDown from '~/components/atoms/form/DropDown'
import Switch from '~/components/atoms/form/Switch'

interface IProps {
  catConf: IParsedCategoryConfiguration
  onChange?: (name: string, value: boolean | string) => void | Promise<void>
  sortOptions: IOptions<string>
  category: ICategory
}

function Merchandize({
  catConf,
  onChange,
  sortOptions,
  category,
}: IProps): JSX.Element {
  const { t } = useTranslation('categories')
  const bundles = useAppSelector(selectBundles)
  const catConfResource = useResource('CategoryConfiguration')
  const fieldConfigMap = new Map(
    catConfResource.supportedProperty.map((field) => [
      field.title,
      getFieldConfig(field),
    ])
  )

  function handleChange(value: string | boolean, event: SyntheticEvent): void {
    onChange((event.target as HTMLInputElement).name, value)
  }

  const isRootCategory = category.level === 1

  return (
    <Paper variant="outlined" style={{ height: 'auto', padding: '22px' }}>
      <Grid
        container
        justifyContent="flex-start"
        alignItems="flex-start"
        spacing={2}
      >
        <Grid item xs={6}>
          <Switch
            checked={catConf?.useNameInProductSearch ?? false}
            infoTooltip={t('name.tooltip')}
            label={t('name')}
            name="useNameInProductSearch"
            onChange={handleChange}
            {...getFieldState(
              catConf as unknown as Record<string, unknown>,
              fieldConfigMap.get('useNameInProductSearch')?.depends
            )}
          />
        </Grid>
        <Grid item xs={6}>
          <DropDown
            infoTooltip={t('select.tooltip')}
            label="Default sorting"
            name="defaultSorting"
            onChange={handleChange}
            options={sortOptions}
            value={catConf?.defaultSorting ?? ''}
            required
            {...getFieldState(
              catConf as unknown as Record<string, unknown>,
              fieldConfigMap.get('defaultSorting')?.depends
            )}
          />
        </Grid>
        {isVirtualCategoryEnabled(bundles) && (
          <Grid item xs={6}>
            <Switch
              checked={catConf?.isVirtual ?? false}
              infoTooltip={t('virtual.tooltip')}
              label={t('virtual')}
              name="isVirtual"
              disabled={isRootCategory}
              onChange={handleChange}
              {...getFieldState(
                catConf as unknown as Record<string, unknown>,
                fieldConfigMap.get('isVirtual')?.depends
              )}
            />
          </Grid>
        )}
      </Grid>
    </Paper>
  )
}

export default Merchandize
