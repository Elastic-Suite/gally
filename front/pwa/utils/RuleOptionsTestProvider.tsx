import { ReactNode } from 'react'

import ruleEngineOperator from '~/public/mocks/rule_engine_operators.json'
import sourceFields from '~/public/mocks/source_fields.json'
import sourceFieldLabels from '~/public/mocks/source_field_labels.json'
import { RuleAttributeType } from 'shared'

import RuleOptionsProvider from '~/components/stateful-providers/RuleOptionsProvider/RuleOptionsProvider'

interface IProps {
  children: ReactNode
}

const sourceFieldLabelsMap = new Map(
  sourceFieldLabels['hydra:member'].map((label) => [
    label.sourceField,
    label.label,
  ])
)

const fields = sourceFields['hydra:member'].map((field) => {
  const label = sourceFieldLabelsMap.get(`/source_fields/${field.id}`)
  return {
    id: field.id,
    code: field.code,
    label: (label || field.defaultLabel) ?? '',
    type: field.type as RuleAttributeType,
  }
})

function RuleOptionsTestProvider(props: IProps): JSX.Element {
  const { children } = props
  return (
    <RuleOptionsProvider
      catalogId={-1}
      localizedCatalogId={-1}
      fields={fields}
      ruleOperators={ruleEngineOperator}
    >
      {children}
    </RuleOptionsProvider>
  )
}

export default RuleOptionsTestProvider
