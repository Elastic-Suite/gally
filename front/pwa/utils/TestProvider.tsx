import { ReactNode } from 'react'

import { schemaContext, userContext } from '~/contexts'
import { IApi, IUser } from 'shared'

import OptionsProvider from '~/components/stateful-providers/OptionsProvider/OptionsProvider'

interface IProps {
  api: IApi
  children: ReactNode
}

const timestamp = Math.floor(Date.now() / 1000)
const expires = 8 * 60 * 60 // 8 hours

const user = {
  iat: timestamp,
  exp: timestamp + expires,
  roles: ['ROLE_ADMIN', 'ROLE_CONTRIBUTOR'],
  username: 'admin@example.com',
} as IUser

function TestProvider(props: IProps): JSX.Element {
  const { api, children } = props
  return (
    <userContext.Provider value={user}>
      <schemaContext.Provider value={api}>
        <OptionsProvider>{children}</OptionsProvider>
      </schemaContext.Provider>
    </userContext.Provider>
  )
}

export default TestProvider
