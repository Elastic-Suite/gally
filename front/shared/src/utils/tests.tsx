import { RenderOptions, render, renderHook } from '@testing-library/react'
import { ReactNode } from 'react'

import { schemaContext } from '../contexts'
import { api } from '../mocks'

export type IExtendedRenderOptions = Omit<RenderOptions, 'queries'>

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  renderOptions: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <schemaContext.Provider value={api}>{children}</schemaContext.Provider>
    )
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions })
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  renderOptions: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <schemaContext.Provider value={api}>{children}</schemaContext.Provider>
    )
  }
  return renderHook(hook, { wrapper: Wrapper, ...renderOptions })
}
