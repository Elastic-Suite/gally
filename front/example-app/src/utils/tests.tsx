import { RenderOptions, render, renderHook } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { ReactNode } from 'react'
import { api } from 'gally-admin-shared'

import TestProvider from './TestProvider'

export type IExtendedRenderOptions = Omit<RenderOptions, 'queries'>

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  renderOptions: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <MemoryRouter>
        <TestProvider api={api}>{children}</TestProvider>
      </MemoryRouter>
    )
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions })
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  renderOptions: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <MemoryRouter>
        <TestProvider api={api}>{children}</TestProvider>
      </MemoryRouter>
    )
  }
  return renderHook(hook, { wrapper: Wrapper, ...renderOptions })
}
