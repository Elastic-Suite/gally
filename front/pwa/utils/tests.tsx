import { RenderOptions, render, renderHook } from '@testing-library/react'
import { PreloadedState } from '@reduxjs/toolkit'
import { ReactNode } from 'react'
import { Bundle, IUser, api } from 'shared'

import { AppStore, RootState, setupStore } from '~/store'

import AppProvider from '~/components/stateful-providers/AppProvider/AppProvider'

import TestProvider from './TestProvider'

export interface IExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export interface IRenderOutput {
  store: AppStore
}

const timestamp = Math.floor(Date.now() / 1000)
const expires = 8 * 60 * 60 // 8 hours

const user = {
  iat: timestamp,
  exp: timestamp + expires,
  roles: ['ROLE_ADMIN', 'ROLE_CONTRIBUTOR'],
  username: 'admin@example.com',
} as IUser

const defaultState = {
  data: {
    api,
    bundles: [Bundle.VIRTUAL_CATEGORY],
    configurations: ['https://localhost/products'],
  },
  user: {
    requestedPath: '',
    token:
      'eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpYXQiOjE2NTk5NTM1NDQsImV4cCI6MTY1OTk4MjM0NCwicm9sZXMiOlsiUk9MRV9BRE1JTiIsIlJPTEVfQ09OVFJJQlVUT1IiXSwidXNlcm5hbWUiOiJhZG1pbkBleGFtcGxlLmNvbSJ9.ZNE2SHmtm9WgWEMmKGMjpcfEym2ysxC_CVcLtzlpuT3jlT0Zyz4VjoQHdTav9I7s9b2QzpeDNt3Sfrbka8IHcVb_TvZle_vII-TkQGal37RtXxKk_EMMQQlzLRO3zDWYy2I0ZPXv6iMlaSWOya4egOG2s1_tzFf4MnNNm2zEQbjzAtZ_Ij9UBMlSxK19AHvHlKU9AelA-NepjUdl6f4GxYPnyIwZw3PS294Wtffj7Kzw1ZZywDljl1xN_flg70cuO0GGkUqBNHrf7zBjPmXbAZZDAefoyQfVqB2v4GzDJfO847XymvBlHtCCjWx0ThK5afyOGL2MAAZkl-AoNjMcjA',
    user,
  },
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...renderHook(hook, { wrapper: Wrapper, ...renderOptions }) }
}
