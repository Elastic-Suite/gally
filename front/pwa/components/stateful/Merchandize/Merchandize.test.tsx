import Merchandize from './Merchandize'
import { renderWithProviders } from '~/utils/tests'

describe('Merchandize match snapshot', () => {
  it('testSelect', () => {
    const first = true
    const sec = true
    const val = 10

    const { container } = renderWithProviders(
      <Merchandize
        virtualCategoryValue={first}
        categoryNameValue={sec}
        sortOptions={[
          { label: 'Position', value: '10' },
          { label: 'Product Name', value: '20' },
          { label: 'Price', value: '30' },
          { label: 'Performance', value: '40' },
        ]}
        sortValue={val}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
