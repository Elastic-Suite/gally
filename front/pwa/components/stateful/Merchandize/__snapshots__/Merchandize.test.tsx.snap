// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merchandize match snapshot testSelect 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded css-11hpgp0-MuiPaper-root"
    style="height: auto; padding: 22px;"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-flggza-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-12mfi9j-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary css-1ghtd4k-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
          >
            name
            <span
              aria-label="name.tooltip"
              class=""
              data-mui-internal-clone-element="true"
              style="display: inline-block; padding-left: 10px; margin-bottom: -5px;"
            >
              <ion-icon
                name="information-circle-outline"
                style="color: rgb(129, 135, 185); width: 20px; height: 20px; display: flex;"
              />
            </span>
          </label>
          <span
            class="MuiSwitch-root MuiSwitch-sizeMedium css-44y1r7-MuiSwitch-root"
          >
            <span
              class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked Mui-checked css-1vk7jze-MuiButtonBase-root-MuiSwitch-switchBase"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                name="useNameInProductSearch"
                type="checkbox"
              />
              <span
                class="MuiSwitch-thumb css-epwt2k-MuiSwitch-thumb"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track css-1mr4grv-MuiSwitch-track"
            />
          </span>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root css-1c9gikl-MuiFormControl-root"
        >
          <div
            class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1qqsdnr-MuiAutocomplete-root"
          >
            <div
              class="MuiFormControl-root css-1pl6002-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required css-ssr94h-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for=":r1:"
              >
                Default sorting
                <span
                  aria-label="select.tooltip"
                  class=""
                  data-mui-internal-clone-element="true"
                  style="display: inline-block; padding-left: 10px; margin-bottom: -5px;"
                >
                  <ion-icon
                    name="information-circle-outline"
                    style="color: rgb(129, 135, 185); width: 20px; height: 20px; display: flex;"
                  />
                </span>
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ufxisk-MuiFormLabel-asterisk"
                >
                  â€‰
                  *
                </span>
              </label>
              <div
                class="InputText__Wrapper css-newnfq"
              >
                <div
                  class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-w8c12a-MuiInputBase-root"
                  required=""
                >
                  <input
                    aria-autocomplete="list"
                    aria-expanded="false"
                    aria-invalid="false"
                    autocapitalize="none"
                    autocomplete="off"
                    class="MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-yz9k0d-MuiInputBase-input"
                    id=":r1:"
                    name="defaultSorting"
                    required=""
                    role="combobox"
                    spellcheck="false"
                    type="text"
                    value="Position"
                  />
                  <div
                    class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                  >
                    <button
                      aria-label="form.open"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                      tabindex="-1"
                      title="form.open"
                      type="button"
                    >
                      <ion-icon
                        name="chevron-down"
                      />
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-12mfi9j-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary css-1ghtd4k-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
          >
            virtual
            <span
              aria-label="virtual.tooltip"
              class=""
              data-mui-internal-clone-element="true"
              style="display: inline-block; padding-left: 10px; margin-bottom: -5px;"
            >
              <ion-icon
                name="information-circle-outline"
                style="color: rgb(129, 135, 185); width: 20px; height: 20px; display: flex;"
              />
            </span>
          </label>
          <span
            class="MuiSwitch-root MuiSwitch-sizeMedium css-44y1r7-MuiSwitch-root"
          >
            <span
              aria-disabled="true"
              class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked Mui-disabled Mui-disabled PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked Mui-disabled Mui-checked Mui-disabled css-1vk7jze-MuiButtonBase-root-MuiSwitch-switchBase"
              tabindex="-1"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                disabled=""
                name="isVirtual"
                type="checkbox"
              />
              <span
                class="MuiSwitch-thumb css-epwt2k-MuiSwitch-thumb"
              />
            </span>
            <span
              class="MuiSwitch-track css-1mr4grv-MuiSwitch-track"
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
